graylog2-web-interface/src/views/components/searchbar/bookmark/BookmarkControls.jsx

```diff
@@ -0,0 +1,192 @@
+// @flow strict
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import { Button, ButtonGroup } from 'react-bootstrap';
+import ViewLoaderContext from 'views/logic/ViewLoaderContext';
+import type { ViewLoaderContextType } from 'views/logic/ViewLoaderContext';
+import { ViewManagementActions } from 'views/stores/ViewManagementStore';
+import UserNotification from 'util/UserNotification';
+import { ViewStore, ViewActions } from 'views/stores/ViewStore';
+import View from 'views/logic/views/View';
+import type { ViewStoreState } from 'views/stores/ViewStore';
+import connect from 'stores/connect';
+
+import BookmarkForm from './BookmarkForm';
+import BookmarkList from './BookmarkList';
+import styles from './BookmarkControls.css';
+
+type Props = {
+  viewStoreState: ViewStoreState,
+};
+
+type State = {
+  showForm: boolean,
+  showList: boolean,
+  newTitle: string,
+};
+
+class BookmarkControls extends React.Component<Props, State> {
+  static contextType = ViewLoaderContext;
+
+  static propTypes = {
+    viewStoreState: PropTypes.object.isRequired,
+  };
+
+  formTarget: any;
+
+  constructor(props: Props, context: ViewLoaderContextType) {
+    super(props, context);
+
+    const { viewStoreState } = props;
+    const { view } = viewStoreState;
+
+    this.state = {
+      showForm: false,
+      showList: false,
+      newTitle: (view && view.title) || '',
+    };
+  }
+
+  toggleFormModal = () => {
+    const { showForm } = this.state;
+    this.setState({ showForm: !showForm });
+  };
+
+  toggleListModal = () => {
+    const { showList } = this.state;
+    this.setState({ showList: !showList });
+  };
+
+  // eslint-disable-next-line no-undef
+  onChangeTitle = (e: SyntheticInputEvent<HTMLInputElement>) => {
+    this.setState({ newTitle: e.target.value });
+  };
+
+  saveSearch = () => {
+    const { newTitle } = this.state;
+    const { viewStoreState } = this.props;
+    const { view } = viewStoreState;
+
+    if (!view.id) {
+      return;
+    }
+
+    const newView = view.toBuilder()
+      .title(newTitle)
+      .type(View.Type.Search)
+      .build();
+
+    ViewManagementActions.update(newView)
+      .then(this.toggleFormModal)
+      .then(() => UserNotification.success(`Saving view "${newView.title}" was successful!`, 'Success!'))
+      .catch(error => UserNotification.error(`Saving view failed: ${this._extractErrorMessage(error)}`, 'Error!'));
+  };
+
+  _extractErrorMessage = (error) => {
+    return (error
+      && error.additional
+      && error.additional.body
+      && error.additional.body.message) ? error.additional.body.message : error;
```

Comment:
I was thinking about how to make this nicer. lodash offers `get`, but using a string for the path breaks type safety and refactorings. Maybe something like this:

```
const orDefault = (fn, dflt) => { try { return fn(); } catch (_) { return dflt; }; };
```

it would be

```
return orDefault(() => error.additional.body.message, error);
```
then.